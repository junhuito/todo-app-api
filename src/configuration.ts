import { DynamicModule, ForwardReference, Type } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { TypeOrmModule, TypeOrmModuleOptions } from '@nestjs/typeorm';
import { EntityClassOrSchema } from '@nestjs/typeorm/dist/interfaces/entity-class-or-schema.type';
import { User } from './user/entities/user.entity';

type DBConnect = (
  | Type<any>
  | DynamicModule
  | Promise<DynamicModule>
  | ForwardReference<any>
)[];

const useEnvironment = (extension: string): string => {
  if (extension === 'test') return 'dev';
  return extension;
};

export const CONFIG_MODULE_ROOT: ConfigModule = ConfigModule.forRoot({
  envFilePath: `${process.cwd()}/.env.${useEnvironment(process.env.NODE_ENV)}`,
  isGlobal: true,
});

export const TYPE_ORM_MODULE_ROOT: TypeOrmModule = TypeOrmModule.forRootAsync({
  imports: [ConfigModule],
  useFactory: (config: ConfigService): TypeOrmModuleOptions => ({
    type: 'postgres',
    host: config.get<string>('DB_HOST'),
    port: Number(config.get<number>('DB_PORT')),
    username: config.get<string>('DB_USERNAME'),
    password: config.get<string>('DB_PASSWORD'),
    database: config.get<string>('DB_NAME'),
    entities: [User],
    synchronize: true,
  }),
  inject: [ConfigService],
});

export const createRootConnection = (): DBConnect => {
  return [CONFIG_MODULE_ROOT, TYPE_ORM_MODULE_ROOT] as DBConnect;
};

export const createDBConnection = (...modules: Type<any>[]): DBConnect => {
  const modulesFlat = modules.flat();
  return [...createRootConnection(), ...modulesFlat];
};

export const createTestingConnection = (
  ...entities: EntityClassOrSchema[]
): DBConnect => {
  const entitiesFlat = entities.flat();
  return [...createRootConnection(), TypeOrmModule.forFeature(entitiesFlat)];
};
