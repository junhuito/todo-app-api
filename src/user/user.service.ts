import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { User } from './entities/user.entity';

@Injectable()
export class UserService {
  constructor(
    @InjectRepository(User) private readonly userRepository: Repository<User>,
  ) {}

  async create(createUserDto: CreateUserDto) {
    console.log('This action adds a new user');
    const newUser = this.userRepository.create(createUserDto);
    return await this.userRepository.save(newUser);
  }

  async findAll() {
    return await this.userRepository.find();
  }

  async findOne(id: number) {
    return await this.userRepository.findOne({ where: { id } });
  }

  async update(id: number, updateUserDto: UpdateUserDto) {
    return await this.userRepository.update(id, { ...updateUserDto });
  }

  async remove(id: number) {
    const user = await this.findOne(id);
    user && this.userRepository.remove(user);
    return 'success';
  }
}
