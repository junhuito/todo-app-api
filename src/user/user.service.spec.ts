import { Test, TestingModule } from '@nestjs/testing';
import { createTestingConnection } from '../configuration';
import { User } from './entities/user.entity';
import { UserService } from './user.service';

describe('UserService', () => {
  let service: UserService;
  let module: TestingModule;

  beforeEach(async () => {
    module = await Test.createTestingModule({
      imports: createTestingConnection(User),
      providers: [UserService],
    }).compile();

    service = module.get<UserService>(UserService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should be get all users', async () => {
    const user = await service.findAll();
    expect(user).toMatchSnapshot();
  });

  afterAll(async () => {
    module?.close();
  });
});
