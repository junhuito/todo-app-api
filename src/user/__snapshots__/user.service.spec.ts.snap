// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService should be get all users 1`] = `
[
  User {
    "email": "",
    "id": "1",
    "password": "",
    "username": "",
  },
  User {
    "email": "",
    "id": "2",
    "password": "",
    "username": "",
  },
  User {
    "email": "",
    "id": "3",
    "password": "",
    "username": "user3",
  },
  User {
    "email": "sunny5771@hotmail.com",
    "id": "4",
    "password": "Junhui8888",
    "username": "user4",
  },
]
`;
